lib/linked_list.rb

```diff
@@ -12,120 +12,249 @@ def initialize(value, next_node = nil)
-    # returns true if a cycle is found, false otherwise.
-    def has_cycle
-      raise NotImplementedError
+    return current.data
+  end
+  
+  ##########################################
+  ##########################################
+  
+  #test with empty list, less than 5 notes, exactly 5 notes
+  def del_five
+    5.times do |i|
+      if @head.nil?
+        return
+      else
+        @head = @head.next
+      end
     end
-
-
-    # Additional Exercises 
-    # returns the value in the first node
-    # returns nil if the list is empty
-    def get_first
-      raise NotImplementedError
+  end
+  
+  def reverse 
+    return if head.nil?
+    current = @head
+    previous = nil
+    
+    until current.next.nil?
+      temp = current.next
+      current.next = previous
+      previous = current
+      current = temp
     end
-
-    # method that inserts a given value as a new last node in the linked list
-    def add_last(value)
-      raise NotImplementedError
+    
+    curent.next = previous
+    @head = current
+  end
+  
+  #go the length and then subtract
+  # def kth_from_last(k)
+  #   return if head.nil?
+  #   current = @head
+  #   k_ahead = @head
+  #   length = 0
+  
+  #   until current.nil?
+  #     length += 1
+  #     current = current.next
+  #   end
+  
+  #   return nil if k > length
+  
+  #   current = @head
+  #   (length - k).times do |i|
+  #     current = current.next
+  #   end
+  
+  #   return current.data
+  # end
+  
+  def kth_from_last(k)
+    current = @head
+    k_ahead = @head
+    
+    k.times do
+      return if k_ahead.nil?
+      k_ahead = k_ahead.next
     end
-
-    # method that returns the value of the last node in the linked list
-    # returns nil if the linked list is empty
-    def get_last
-      raise NotImplementedError
+    
+    until k_ahead.nil?
+      current = current.next
+      k_ahead = k_ahead.next
     end
-
-    # method to insert a new node with specific data value, assuming the linked
-    # list is sorted in ascending order
-    def insert_ascending(value)
-      raise NotImplementedError
+    
+    return current.data
+  end
+  
+  # method to find if the linked list contains a node with specified value
+  # returns true if found, false otherwise
+  def search(value)
+    current = @head
+    
+    until current == nil
+      return true if current.data == value
+      current = current.next
     end
-
-    # Helper method for tests
-    # Creates a cycle in the linked list for testing purposes
-    # Assumes the linked list has at least one node
-    def create_cycle
-      return if @head == nil # don't do anything if the linked list is empty
-
-      # navigate to last node
-      current = @head
-      while current.next != nil
-          current = current.next
+    
+    return false
+  end
+  
+  # method to return the max value in the linked list
+  # returns the data value and not the node
+  def find_max
+    current = @head
+    max = 0

```

Comment:
What if it's a list of negative numbers

```suggestion
    max = @head.nil? ? nil : @head.data
```